job/jobcache.go

```diff
@@ -1,36 +1,88 @@
 package job
 
 import (
+	"os"
+	"os/signal"
 	"sync"
+	"time"
 )
 
-var (
-	AllJobs = &JobCache{
-		Jobs:   make(map[string]*Job),
-		rwLock: sync.Mutex{},
+type JobCache interface {
+	Get(id string) *Job
+	GetAll() map[string]*Job
+	Set(j *Job)
+	Delete(id string)
+	Persist() error
+}
+
+type MemoryJobCache struct {
+	// Jobs is a map from Job id's to pointers to the jobs.
+	// Used as the main "data store" within this cache implementation.
+	jobs            map[string]*Job
+	rwLock          sync.Mutex
+	jobDB           JobDB
+	persistWaitTime time.Duration
+}
+
+func NewMemoryJobCache(jobDB JobDB, persistWaitTime time.Duration) *MemoryJobCache {
+	if persistWaitTime == 0 {
+		persistWaitTime = 5 * time.Second
 	}
-)
+	return &MemoryJobCache{
+		jobs:            map[string]*Job{},
+		rwLock:          sync.Mutex{},
+		jobDB:           jobDB,
+		persistWaitTime: persistWaitTime,
+	}
+}
+
+func (c *MemoryJobCache) Start() {
+	// Prep cache
+	allJobs, err := c.jobDB.GetAll()
+	if err != nil {
+		log.Fatal(err)
+	}
+	for _, j := range allJobs {
+		j.StartWaiting(c)
+		c.Set(j)
+	}
+
+	// Occasionally, save items in cache to db.
+	go c.PersistEvery()
+
+	// Process-level defer for shutting down the db.
+	ch := make(chan os.Signal)
+	signal.Notify(ch, os.Interrupt, os.Kill)
+	go func() {
+		s := <-ch
+		log.Info("Process got signal: %s", s)
+		log.Info("Shutting down....")
 
-type JobCache struct {
-	Jobs   map[string]*Job
-	rwLock sync.Mutex
+		// Persist all jobs to database
+		c.Persist()
+
+		// Close the database
+		c.jobDB.Close()
+
+		os.Exit(1)
+	}()
 }
 
-func (c *JobCache) Get(id string) *Job {
+func (c *MemoryJobCache) Get(id string) *Job {
 	c.rwLock.Lock()
 	defer c.rwLock.Unlock()
 
-	return c.Jobs[id]
+	return c.jobs[id]
 }
 
-func (c *JobCache) GetAll() map[string]*Job {
+func (c *MemoryJobCache) GetAll() map[string]*Job {
 	c.rwLock.Lock()
 	defer c.rwLock.Unlock()
 
-	return c.Jobs
+	return c.jobs
```

Comment:
Instead of passing around a map, pass around a custom type that's a struct
{ map[T1]T2; sync.RWMutex } and have the caller call .Lock(), defer
.Unlock() (or .RLock(), defer .RUnlock(), depending on whether the caller
is modifying the map or just reading through it).

The Syndicate Pro API does this all over the place, so check that out for
examples if you need to.
On Aug 12, 2015 7:43 PM, "AJ Bahnken" <notifications@github.com> wrote:

> In job/jobcache.go
> <https://github.com/ajvb/kala/pull/36#discussion_r36936654>:
>
> >  	c.rwLock.Lock()
> >  	defer c.rwLock.Unlock()
> >
> > -	return c.Jobs
> > +	return c.jobs
>
> @elimisteve <https://github.com/elimisteve> Last weekend you mentioned a
> solution for this? Mind describing it too me? I'm now in a bit more of a
> rush to get this "production-ready" :P
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ajvb/kala/pull/36/files#r36936654>.
>